### ISSUE TITLE
CosmosDBTrigger -- support single Document for binding

### ISSUE THREAD

**Issue by brettsam** [Azure/azure-webjobs-sdk-extensions#294]
Right now we only support `IReadOnlyList<Document>` as an input type. That makes it tough to bind to anything at runtime b/c you could have many documents, which means that we don't know which one to look at for your binding values (i.e. if you want to take a changed document and do an auto-lookup for related documents using a CosmosDB input binding).

Right now you have to either move to multiple functions joined by a queue or do the work yourself with DocumentClient internally. If we could support `Document` by itself as a trigger object somehow, that'd open up the possibility to have richer runtime binding here.

**Comment by mikhailshilkov** [comment]
It would also be nice to support single **typed** document, i.e. with a custom C# class.

**Comment by Saphirim** [comment]
Would be a great addition to further evolve the CosmosDB -> FunctionApps workflow

**Comment by brettsam** [comment]
@mikhailshilkov -- I split your comment out into a separate issue (#389).

@ealsur -- would setting `MaxItemsPerInvocation` allow us to restrict this to support a single document?

**Comment by ealsur** [comment]
@brettsam Yes it should, you would still receive a List/Array of documents but it would contain 1 element.

Having said that, it might not be the best approach. The Trigger is optimized for a stream of changes scenario, reducing the batch size to 1 it would certainly made it so it needs to read changes much more frequently (if there are 50 pending changes it needs to do 50 reads instead of 1). If you need to trigger a Function based on 1 document, why not send the batch items to a Queue after they are received in batch?

**Comment by brettsam** [comment]
That's similar to what we do for blobs already -- so maybe that's the right approach. When we find any blobs, we stick the blob id into a queue message and then submit it. That allows for broader scale-out as well as you don't need many readers. For blobs, only one instance actually checks for new blobs, but many can read from the queue.

**Comment by paulbatum** [comment]
Yeah I'm not sure we should add this capability. Cosmos DB and Event Hubs have a lot in common when it comes to performance, and my findings for Event Hubs were that it is massively slower (and in the consumption plan, more expensive!) to write functions that receive a single EventData rather than an array. I've had a number of support cases that ended up being root caused to this issue. I'm considering a breaking change in V2 that would force event hubs to only support batched mode.

As @ealsur points out, a user can work around this by introducing a queue.  This will work fine in low throughput scenarios. In high throughput scenarios, the developer really needs to do their processing in batches.

**Comment by mikhailshilkov** [comment]
Single-document trigger could have nice semantics in combination with input/output bindings and templates. Looks very cool for demos, at least :)
I'm not sure that every user of Cosmos DB cares about performance. I used it for collections of several thousands documents, so I couldn't care less about batching the changes.

**Comment by paulbatum** [comment]
I agree performance is not always a priority! But we have been burned by shipping features that do not perform/scale well (for example our experimental python support, which calls python.exe for every invocation), so we're becoming more careful about this, especially when there are workarounds that can be used to unblock the scenario - in this case, writing the batch to a queue and using a queue triggered function would give you that ability to use input bindings.

**Comment by mikhailshilkov** [comment]
Speaking of Storage Queue and Service Bus triggers, it would then make sense to add batching capabilities to those too.
As it is now, the triggers' capabilities are quite inconsistent.

**Comment by ealsur** [comment]
I understand what you mean, a couple of years ago I created this extension that works with the Queue trigger in batches: https://github.com/ealsur/WebJobs.Extensions.GroupQueueTrigger 

Maybe one day it's useful enough to be integrated into the Functions runtime 😄

**Comment by Saphirim** [comment]
All Queue-/EventHub-/xy-alternatives lack the possibility to edit a new/updated ComsmosDB-document "in place" and just put it back into the storage by using an output-binding, like function apps support for other integrations.

This is especially interesting, since the CosmosDb-changefeed does not include information about the C(R)UD-operation processed on the item in the changefeed. So the information about a document "being new" has to be stored within the custom data and update after the first run through my function app, so that for consecutive invocations after updating the document the function can recognize it as "only updated" and "not new". The need to persist this document state in my custom data from within the function app today requires a CosmosDb client updating the specific document. An simple output binding for (single) documents combined with an input binding would reduce complexity here drastically.

Correct me if I'm wrong, I'd be very thankful for a nicer pattern of achieving this!

**Comment by bassambst** [comment]
2019 and this is still an issue!
Azure team published a tutorial about this as it does really work!
https://docs.microsoft.com/en-us/learn/modules/chain-azure-functions-data-using-bindings/5-read-data-with-input-bindings-portal-lab

**Comment by christian-vorhemus** [comment]
Is there any news on this? We are working on a scenario where we listen on changes of a CosmosDB collection and when triggered, we need additional "reference data" from another collection based on some parameters of the document we got from the cosmosDBTrigger binding.

The improvement you described @brettsam would perfectly fit and we would be able to avoid an additional call to Cosmos in the code.

**Comment by paulbatum** [comment]
As discussed above, we have concerns with introducing this option from a performance perspective. Let me build on the argument a little further using your scenario. 

Lets imagine for a moment that we implemented this feature and you used an input binding to retrieve one additional set of reference data. This means for every change in the change feed, you'd be making a call to cosmos to get the reference data. If we read 100 items from the change feed, that's 100 calls.

Now lets say you're going to implement this scenario today, and you don't have this feature. You're going to be passed a collection of changes, say with a 100 items in them. You could iterate through those items one by one sequentially and you'd end up with the same number of extra calls as above (100).

But you can be smarter. You could:
a) Make those 100 calls to cosmos in parallel. Or if that hammers your cosmos instance too hard, you could make 10 batches of 10 concurrent calls each. Whatever you choose is going to be faster than retrieving them one at a time.
b) Construct a single cosmos query that is able to return all of the reference data that is applicable to those 100 items. Now you've managed to put less load on your database and made your function run faster!

This example does not capture all of the performance concerns we have with single document mode, but hopefully its at least illustrative. Batching performs better naturally AND it offers opportunity for additional optimizations.
